File: Question 7/valid_parenthesis.py
class Solution(object):
    def isValid(self, s):
        stack = [] 
        for c in s: 
            if c in '([{': # if the character is an opening bracket
                stack.append(c) 
            else: # if the character is a closing bracket
                if not stack or \
                    (c == ')' and stack[-1] != '(') or \
                    (c == '}' and stack[-1] != '{') or \
                    (c == ']' and stack[-1] != '['):
                    return False # the string is not valid, so return false
                stack.pop()
        return not stack 

File: Question 7/valid_parenthesis.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> st;
        int i;
        char ch;
        for(i=0;i<s.length();i++)
        {
            ch=s[i];

            if(s[i]=='{'||s[i]=='('||s[i]=='[')
               st.push(ch);

            else
            {
                if(!st.empty())
                {
                   char top=st.top();
                    if(ch==')'&&top=='('||ch=='}'&&top=='{'||ch==']'&&top=='[')
                    st.pop();
                    else
                    return false;
                }
                else
                return false;
            }
        }
                if(st.empty())
                return true;
                return false;
            }
        
    
};
